File: backend/schemas/error_book.py
"""板块二错题本 schema（对齐 openapi.yaml v1.5 错题本段）。"""
from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class ErrorRecordCreate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    subject: str
    raw_text: str = Field(..., alias="rawText")
    student_answer: str | None = Field(None, alias="studentAnswer")
    correct_answer: str | None = Field(None, alias="correctAnswer")
    error_type: str | None = Field(None, alias="errorType")
    error_note: str | None = Field(None, alias="errorNote")
    point_ids: List[str] = Field([], alias="pointIds")


class ErrorRecordUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    error_type: str | None = Field(None, alias="errorType")
    error_note: str | None = Field(None, alias="errorNote", max_length=4000)
    status: str | None = None
    point_ids: List[str] | None = Field(None, alias="pointIds")


class LinkedPoint(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    point_id: str = Field(..., alias="pointId")
    name: str | None = None
    confidence: float | None = None


class ErrorRecord(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    error_id: str = Field(..., alias="errorId")
    subject: str
    raw_text: str = Field(..., alias="rawText")
    student_answer: str | None = Field(None, alias="studentAnswer")
    correct_answer: str | None = Field(None, alias="correctAnswer")
    error_type: str | None = Field(None, alias="errorType")
    error_note: str | None = Field(None, alias="errorNote")
    status: str
    points: List[LinkedPoint] = []
    created_at: str = Field(..., alias="createdAt")
    last_reviewed_at: str | None = Field(None, alias="lastReviewedAt")


class ErrorBookList(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    items: List[ErrorRecord]
    pagination: dict


class ErrorRecordDeleted(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    deleted: bool
    error_id: str = Field(..., alias="errorId")


class ReviewSubmit(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    recall_correct: bool = Field(..., alias="recallCorrect")


class ReviewResult(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    correct: bool
    next_review_at: str = Field(..., alias="nextReviewAt")
    interval_days: int = Field(..., alias="intervalDays")

File: frontend/src/services/errorBook.ts
/**
 * 板块二 · 错题本服务（v2.1 转正）
 *
 * 对接真实后端 /api/v1/error-book/*（docs/openapi.yaml v1.5 错题本段）。
 * 原文 rawText/studentAnswer/correctAnswer/errorNote 只在本地，永不出域。
 */

import { apiDelete, apiGet, apiPatch, apiPost } from './http';

export interface LinkedPoint {
  pointId: string;
  name?: string | null;
  confidence?: number | null;
}

export interface ErrorRecord {
  errorId: string;
  subject: string;
  rawText: string;
  studentAnswer?: string | null;
  correctAnswer?: string | null;
  errorType?: string | null;
  errorNote?: string | null;
  status: 'open' | 'resolved';
  points: LinkedPoint[];
  createdAt: string;
  lastReviewedAt?: string | null;
}

export interface ErrorRecordCreate {
  subject: string;
  rawText: string;
  studentAnswer?: string;
  correctAnswer?: string;
  errorType?: string;
  errorNote?: string;
  pointIds?: string[];
}

export interface ErrorRecordUpdate {
  errorType?: string;
  errorNote?: string;
  status?: 'open' | 'resolved';
  pointIds?: string[];
}

export interface ReviewResult {
  correct: boolean;
  nextReviewAt: string;
  intervalDays: number;
}

/** GET /error-book 列表 */
export function fetchErrorBook(params: {
  subject?: string;
  status?: string;
  page?: number;
  pageSize?: number;
}): Promise<{ items: ErrorRecord[]; pagination: { page: number; pageSize: number; total: number } }> {
  return apiGet('/error-book', params as Record<string, string | number | undefined>);
}

/** POST /error-book 录入 */
export function createErrorRecord(payload: ErrorRecordCreate): Promise<ErrorRecord> {
  return apiPost('/error-book', payload);
}

/** GET /error-book/{id} 详情 */
export function fetchErrorRecord(errorId: string): Promise<ErrorRecord> {
  return apiGet(`/error-book/${errorId}`);
}

/** PATCH /error-book/{id} 更新 */
export function updateErrorRecord(errorId: string, payload: ErrorRecordUpdate): Promise<ErrorRecord> {
  return apiPatch(`/error-book/${errorId}`, payload);
}

/** DELETE /error-book/{id} 软删 */
export function deleteErrorRecord(errorId: string): Promise<{ deleted: boolean; errorId: string }> {
  return apiDelete(`/error-book/${errorId}`);
}

/** POST /error-book/{id}/review 复习 */
export function reviewErrorRecord(errorId: string, recallCorrect: boolean): Promise<ReviewResult> {
  return apiPost(`/error-book/${errorId}/review`, { recallCorrect });
}
